Scripts to update config/overleaf.rc and config-seed rebrand

mserranom opened the pull request:
## Description
<!-- Goal of the pull request -->

- Updates variables `config-seed/overleaf.rc` with `OVERLEAF_` prefix, and replaces occurrences of those variables through the repo.
- Updates `config-seed/variables.env`, and updates the documentation accordingly.
- Adds scripts to upgrade users' `config/overleaf.rc` by extending the scripts added in https://github.com/overleaf/toolkit/pull/214

This PR needs to be merged once CE/SP 5.0.0 is shipped,  contains the changes to `config-seed/version` and `CHANGELOG`. The latter needs to be amended with the release date and release note links.


## Related issues / Pull Requests
<!-- Fixes #xyz, Contributes to #xyz, Related to #xyz-->


## Contributor Agreement

- [x] I confirm I have signed the [Contributor License Agreement](https://github.com/overleaf/overleaf/blob/master/CONTRIBUTING.md#contributor-license-agreement)

das7pad commented on CHANGELOG.md:
```diff
@@ -204,5 +212,5 @@
 ## 2020-10-09
 ### Added
-- Add `SHARELATEX_PORT` option to `overleaf.rc` file, which defaults
+- Add `OVERLEAF_PORT` option to `overleaf.rc` file, which defaults
```
This looks like a historic entry that we might want to keep as is?

mserranom replied:
That slipped, thanks!

das7pad commented on lib/docker-compose.base.yml:
```diff
@@ -7,17 +7,17 @@ services:
-          SHARELATEX_MONGO_URL: "${MONGO_URL}"
-          SHARELATEX_REDIS_HOST: "${REDIS_HOST}"
+          OVERLEAF_MONGO_URL: "${MONGO_URL}"
+          OVERLEAF_REDIS_HOST: "${REDIS_HOST}"
```
Hmm, this is breaking 4x setups. I suspect we need to use a dynamic variable here -- or define both (if so, for how many future releases?)?

mserranom replied:
I'm testing different flows. 4.x setups should be prompted to upgrade their `overleaf.rc`, working on it right now.

das7pad commented on bin/doctor:
```diff
@@ -353,6 +353,87 @@ function check_config_files() {
           print_point 2 "SHARELATEX_HISTORY_BACKEND: fs"
         fi
+
+        if [[ "${OVERLEAF_FILESTORE_BACKEND:-fs}" == "s3" ]]; then
```
Do we need this duplicated? The doctor could instead check the major version and complain about SHARELATEX_ entries (major>=5) or OVERLEAF_ entries (major<5).

das7pad commented on bin/doctor:
```diff
@@ -185,11 +185,11 @@ function check_config_files() {
         # Check some vars from the RC file
-        if [[ "${SHARELATEX_DATA_PATH:-null}" != "null" ]]; then
-          print_point 2 "SHARELATEX_DATA_PATH: $SHARELATEX_DATA_PATH"
+        if [[ "${OVERLEAF_DATA_PATH:-null}" != "null" ]]; then
```
4.x setups are still allowed to use the SHARELATEX_ prefixed ones. We could add a fallback like I did in https://github.com/overleaf/toolkit/pull/221

mserranom replied:
After upgrading the toolkit they should have these migrated to `OVERLEAF_` prefix. We can print both, and also print information in the doctor about whether their successfully migrated their `overlearf.rc`.

das7pad reviewed (commented):
Stepping back a bit, maybe it's too ambitious to support both 4.x setups and 5.x setups at the same time? We could require customers to stay on an older Toolkit version to run Server Pro 4.x and upgrade to the latest Toolkit version when they upgrade to Server Pro 5.x.

@briangough @emcsween @mmoreauoverleaf @mlevans0 @mserranom  WDYT?

mserranom commented:
I pushed some changes to check `config/overleaf.rc` without checking `config/version` first and tested some scenarios

**Run an existing `4.x` instance in this branch**

```
✗ bin/up -d
WARNING: some runtime variables defined in config/overleaf.rc
contain 'SHARELATEX_' in their name. The variables should be
renamed to 'OVERLEAF_'.
you can upgrade your config/overleaf.rc by running bin/rename-rc-vars
Do you want to continue? (y/n): n
Exiting.

✗ bin/rename-rc-vars
This script will update your config/overleaf.rc.
We recommend backing up your config with bin/backup-config.
Do you want to continue? (y/n): y
Found 5 lines with SHARELATEX_ in overleaf.rc
Creating backup file config/__old-overleaf.rc.2024.02.13-12.31.10
Replacing 'SHARELATEX_' with 'OVERLEAF_' in config/overleaf.rc
Updated 5 lines in /Users/m.serrano/workspace/toolkit/config/overleaf.rc
Done.

✗ bin/up -d

✗ curl http://localhost
Found. Redirecting to /login%
```

Tested:

- Create edit, compile project
- Navigate through history


**Upgrade from `4.x` to `5.0.0-RC`**

```
✗ cat config/version
4.2.0
✗ echo 5.0.0-RC > config/version
✗ cat config/version
5.0.0-RC

✗ bin/up -d
WARNING: some environment variables defined in config/variables.env
contain 'SHARELATEX_' in their name
Starting with Overleaf CE and Server Pro 5.0.0 the environment variables
use the prefix 'OVERLEAF_'  instead of 'SHARELATEX_'.
Please check your config/version, and config/variables.env files,
you can upgrade your config/variables.env by running bin/rename-env-vars-5-0
Do you want to continue? (y/n): n
Exiting.

✗ bin/rename-env-vars-5-0
This script will update your config/variables.env and config/overleaf.rc.
We recommend backing up your config with bin/backup-config.
Do you want to continue? (y/n): y
Found 60 lines with SHARELATEX_ in variables.env
Creating backup file config/__old-variables.env.2024.02.13-13.06.00
Replacing 'SHARELATEX_' with 'OVERLEAF_' in config/variables.env
Updated 60 lines in /Users/m.serrano/workspace/toolkit/config/variables.env
Done.

✗ bin/up -d

✗ curl http://localhost
Found. Redirecting to /login%
```

mserranom commented:
I've added a couple of changes after a conversation with @das7pad, rebased and worked out a better commit breakdownto facilitate the code review (lost some attribute, apologies).

das7pad commented on bin/doctor:
```diff
@@ -185,11 +187,20 @@ function check_config_files() {
-          print_point 2 "SHARELATEX_DATA_PATH: $SHARELATEX_DATA_PATH"
+        if [[ "$IMAGE_VERSION_MAJOR" -lt 5 ]]; then
+          if [[ "${SHARELATEX_DATA_PATH:-null}" != "null" ]]; then
+            print_point 2 "SHARELATEX_DATA_PATH: SHARELATEX_DATA_PATH"
```
```suggestion
            print_point 2 "SHARELATEX_DATA_PATH: $SHARELATEX_DATA_PATH"
```

das7pad commented on bin/rename-env-vars-5-0:
```diff
@@ -19,9 +19,9 @@ fi
 
 function usage() {
-  echo "Usage: bin/update-env"
+  echo "Usage: bin/update-env-vars-5-0"
```
```suggestion
  echo "Usage: bin/rename-env-vars-5-0"
```

das7pad commented on bin/up:
```diff
@@ -36,17 +36,34 @@ function check_config() {
+  local env_occurrences=$(grep -o "$invalid_prefix" "$TOOLKIT_ROOT/config/variables.env" | wc -l | sed 's/ //g')
+  set -o pipefail
+
+  if [ "$env_occurrences" -gt 0 ]; then
```
Just curious, was there anything "wrong" with the previous exit code based check to identify if any lines need migrating?

mserranom replied:
Yeah, here's an example:

With `pipefail`:

```bash
#! /usr/bin/env bash

set -euo pipefail

touch __tmp

count=$(grep -o SOMETHING __tmp | wc -l | sed 's/ //g')

echo "found $count occurrences"
```

``` bash
> ./example # empty response
> echo $?
1
```

Without `pipefail`:

```bash
#! /usr/bin/env bash

set -euo pipefail

touch __tmp

set +o pipefail
count=$(grep -o SOMETHING __tmp | wc -l | sed 's/ //g')
set +o pipefail

echo "found $count occurrences"
```

``` bash
> ./example # empty response
found 0 occurrences
> echo $?
0
```

das7pad replied:
Right, for the case where you want to know the number of entries we need a bit of larger snippet. But here we only care about "has matching entries" (`grep -q` -> exit 0) or "does not have any matching entries" (`grep -q` -> exit 1). 

Btw, `grep --count` emits `0`, `1` or any other number of entries it found, which might be "cleaner" :) You would still need to handle the non-zero exit code for the case where zero entries were found.

```bash
jackermann@x1:/tmp$  cat foo.log 
foo
jackermann@x1:/tmp$  grep -e foo --count foo.log 
1
jackermann@x1:/tmp$  echo $?
0
jackermann@x1:/tmp$  grep -e bar --count foo.log 
0
jackermann@x1:/tmp$  echo $?
1
```

das7pad commented on bin/rename-rc-vars:
```diff
@@ -0,0 +1,43 @@
+source "$TOOLKIT_ROOT/lib/shared-functions.sh"
+
+function usage() {
+  echo "Usage: bin/rename-rc-vars"
```
```suggestion
  echo "Usage: bin/rename-rc-vars-5-0"
```
Similar to the variables script, it might be worth adding a version suffix here, just in case we need to migrate entries in the future.

mserranom replied:
The `overleaf.rc` is not linked to the version of Server Pro (we don't check the version in `bin/upgrade` before updating it), that's why I didn't add the suffix.

das7pad replied:
Good point!

das7pad commented on lib/docker-compose.base.yml:
```diff
@@ -7,17 +7,19 @@ services:
           REDIS_PORT: "${REDIS_PORT}"
-          SHARELATEX_MONGO_URL: "${MONGO_URL}"
-          SHARELATEX_REDIS_HOST: "${REDIS_HOST}"
+          OVERLEAF_MONGO_URL: "${MONGO_URL}"
+          OVERLEAF_REDIS_HOST: "${REDIS_HOST}"
```
Do we need these two entries still with the lines below in place now?

mserranom replied:
Correct, that's a leftover. Thanks!

das7pad commented on bin/docker-compose:
```diff
@@ -56,21 +56,23 @@ function set_base_vars() {
   export IMAGE="$image_name:$IMAGE_VERSION"
 
-  if [[ ${SHARELATEX_LISTEN_IP:-null} == "null" ]];
+  OVERLEAF_LISTEN_IP=${OVERLEAF_LISTEN_IP:-${SHARELATEX_LISTEN_IP:-}}
```
I think we can remove this fallback (and the port one below) now that we require the rc file to be migrated.

mserranom replied:
Done in https://github.com/overleaf/toolkit/pull/217/commits/0fee1a6f7d42baddf7396d9042017d6eb3cecb68

das7pad commented on lib/default.rc:
```diff
@@ -2,8 +2,8 @@
 SERVER_PRO=false
-SHARELATEX_DATA_PATH=data/sharelatex
-SHARELATEX_PORT=80
+OVERLEAF_DATA_PATH=data/sharelatex
```
cc @briangough WDYT about pointing new setups at `data/overleaf`?

das7pad replied:
Oh, sorry, I meant to comment in the `lib/config-seed/overleaf.rc` :see_no_evil: 

The default.rc is used by all deployments as fallback and should not migrate that one.

mserranom replied:
I'll update `OVERLEAF_DATA_PATH` to `data/overleaf` in `lib/config-seed/overleaf.rc` 👍🏻

mserranom replied:
Done in https://github.com/overleaf/toolkit/pull/217/commits/4f848c3ee31f3cf1519a2ef5532d3e6c46de5706

das7pad reviewed (commented):
Looking great! (Not tested yet)

For users who `git pull` before running `bin/upgrade`, it might be worth adding a check in `bin/docker-compose` to refuse to work when the rc file has not been migrated yet. Otherwise custom `..._DATA_PATH` would not be picked up.

mserranom commented:
> For users who `git pull` before running `bin/upgrade`, it might be worth adding a check in `bin/docker-compose` to refuse to work when the rc file has not been migrated yet. Otherwise custom `..._DATA_PATH` would not be picked up.

That's a good point. I think moving the checks from `bin/up` to `bin/docker-compose` will do the thing 👍🏻

mserranom commented:
> > For users who `git pull` before running `bin/upgrade`, it might be worth adding a check in `bin/docker-compose` to refuse to work when the rc file has not been migrated yet. Otherwise custom `..._DATA_PATH` would not be picked up.
> 
> That's a good point. I think moving the checks from `bin/up` to `bin/docker-compose` will do the thing 👍🏻

Done in https://github.com/overleaf/toolkit/pull/217/commits/eeb9ae3d827039f289b0fc361060b5c868166311

briangough commented on bin/docker-compose:
```diff
@@ -18,6 +18,39 @@ fi
 source "$TOOLKIT_ROOT/lib/shared-functions.sh"
 
+function check_sharelatex_env_vars() {
+  set +o pipefail
+  local rc_occurrences=$(grep -o SHARELATEX_ "$TOOLKIT_ROOT/config/overleaf.rc" | wc -l | sed 's/ //g')
+  set -o pipefail
```
Since the `$(...)` runs in a subshell you can temporarily disable pipefail inside the subshell like this:`local rc_occurences=$(set +o pipefail && grep ...)`. Then you don't need to reenable it afterward.

mserranom replied:
Net! I upgraded the commands in https://github.com/overleaf/toolkit/pull/217/commits/f1a50d986e93d0fb3224b00d255e4a1492b755a5

briangough commented on bin/docker-compose:
```diff
@@ -18,6 +18,39 @@ fi
+    invalid_prefix="OVERLEAF_"
+  fi
+
+  set +o pipefail
+  local env_occurrences=$(grep -o "$invalid_prefix" "$TOOLKIT_ROOT/config/variables.env" | wc -l | sed 's/ //g')
+  set -o pipefail
```
Same pipefail trick as above.

briangough commented on lib/shared-functions.sh:
```diff
@@ -27,18 +27,19 @@ prompt() {
 rebrand_sharelatex_env_variables() {
+  local filename=$1
   set +o pipefail
-  sharelatex_occurrences=$(grep -o "SHARELATEX_" "$TOOLKIT_ROOT/config/variables.env" | wc -l | sed 's/ //g')
+  sharelatex_occurrences=$(grep -o "SHARELATEX_" "$TOOLKIT_ROOT/config/$filename" | wc -l | sed 's/ //g')
   set -o pipefail
```
Another pipefail subshell.

mserranom commented:
It turns `SHARELATEX_MONGO_URL` and `SHARELATEX_REDIS_HOST` where only working in <v4 because I forgot to remove the original variables 🤦🏻 (https://github.com/overleaf/toolkit/pull/217/commits/f67fae9407fa31d1e020be6ee1244f9abedb3589)

I've ended up extracting the variables into override compose files in https://github.com/overleaf/toolkit/pull/217/commits/0dd9750a5870a825abc885329e7401e9602f2a65. Since we already have similar mechanisms in place I found it a better option than running substitutions in `docker-compose.base.yml`. It's tested in `4.2.3` and the nightly build tagged as `5.0.0-RC`.

das7pad reviewed (commented):
The upgrade script needs some more tweaks.

```
Upgrade image? [y/n] y
Upgrading config/version from 4.2.3 to 5.0.0-RC2
Do you want to continue? (y/n): n                                <- from sub-shell
docker services are up, stop them first?
Stop docker services? [y/n]
```

The 2nd prompt here is from the sub-shell in `services_up`, which tries to run `bin/docker-compose stop`, which in turn asks the user to migrate the config.
When rejecting that prompt, the upgrade script takes the output buffer as "things are up" and prompts to stop things -- even if nothing is running.

WDYT about performing the RC migration before running `handle_image_upgrade`?

The `services_up` check could be improved in a separate PR.

mserranom commented:
Rebased and updated with changes from https://github.com/overleaf/internal/pull/17259 (Cc @das7pad)

das7pad commented on bin/doctor:
```diff
@@ -185,11 +187,20 @@ function check_config_files() {
+          else
+            print_point 2 "SHARELATEX_DATA_PATH: MISSING !"
+            add_warning "rc file, SHARELATEX_DATA_PATH not set"
+          fi
```
`bin/doctor` now prints this when the `overleaf.rc` has not been migrated: 86ef323f7068ffc35512749f78fbe0e81f40f91a

```
toolkit$ bin/doctor
====== Overleaf Doctor ======
Rebranding from ShareLaTeX to Overleaf
  The Toolkit has adopted to Overleaf brand for its variables.
  Your config/overleaf.rc still has 3 variables using the previous ShareLaTeX brand.
  Moving forward the 'SHARELATEX_' prefixed variables must be renamed to 'OVERLEAF_'.
  You can migrate your config/overleaf.rc to use the Overleaf brand by running:

    toolkit$ bin/rename-rc-vars

```

And I've cleaned up the lookup of these old ShareLaTeX branded rc entries via aa77274c19450c365b0b6a18c02fc295de59230d

das7pad commented:
@briangough could you take another look at the commits I added today? Thanks!

das7pad commented:
Rebased to fix the conflicts (bonus: I've also switched to `5.0.0-RC1` while fixing the version conflict).

das7pad commented:
Fix for https://github.com/overleaf/internal/issues/17375 https://github.com/overleaf/toolkit/commit/f8d14ed8afb6ced21455203b238fa938d72932be

```
toolkit$ bin/up
Initiating Mongo replica set...
Rebranding from ShareLaTeX to Overleaf
  The 'TEXMFVAR' override is not needed since Server Pro/Overleaf CE version 3.2 (August 2022) and it conflicts with the rebranded paths.
  Please remove the following entry from your config/variables.env:

    TEXMFVAR=/var/lib/sharelatex/tmp/texmf-var
```
